File: NoteS/controllers/PublicCommentsController.cs
using Microsoft.AspNetCore.Mvc;
using NoteS.Attributes;
using NoteS.exceptions;
using NoteS.models.dto;
using NoteS.models.entity;
using NoteS.models.mappers;
using NoteS.services;
using NoteS.tools;
using Swashbuckle.AspNetCore.Annotations;

namespace NoteS.controllers;

[ApiController]
[Route("api/public/{accountName}")]
[ProducesResponseType(StatusCodes.Status403Forbidden)]
[ProducesResponseType(StatusCodes.Status401Unauthorized)]
public class PublicCommentsController(
    AccountRegisterService register,
    CommentInformationService commentInformationService,
    CommentEditService commentEditService,
    UniversalDtoMapper um) : GeneralPreconditionController(register)
{
    [HttpGet("notes/{pathNote}/comments")]
    [KeycloakAuthorize(Policies.READ_COMMENTS)]
    [SwaggerOperation(Description = "Список комментариев к заметке")]
    [ProducesResponseType(StatusCodes.Status200OK)]
    public PageDto<NoteSearchContentResponseDto> Comments([FromQuery] AccName accountName,
        [FromQuery] NotePath pathNote,
        [FromQuery] PaginationRequestDto pagination)
    {
        Check(accountName);
        var comments = commentInformationService.Comments(accountName, pathNote, pagination, pagination);
        return um.OfPage(comments);
    }

    [HttpPost("notes/{pathNote}/comments")]
    [KeycloakAuthorize(Policies.READ_COMMENTS, Policies.READ_NOTES, Policies.CREATE_COMMENTS)]
    [SwaggerOperation(Description = "Создание комментария к заметке")]
    [ProducesResponseType(StatusCodes.Status201Created)]
    public async Task<ActionResult<CommentCreateResponseDto>> CreateComment([FromQuery] AccName accountName,
        [FromQuery] NotePath pathNote,
        [FromBody] CommentCreateRequestDto createDto)
    {
        Check(accountName);
        CommentCreateResponseDto note = await commentEditService.CreateComment(accountName, pathNote, createDto);

        return Created(Url.Action("GetNote", "PublicNote",
            new { accountName.AccountName, pathNote = note.Path }, Request.Scheme), note);
    }

    [HttpPost]
    [KeycloakAuthorize(Policies.READ_COMMENTS, Policies.READ_NOTES, Policies.EDIT_OWN_COMMENTS)]
    [Route("comments/{pathNote}")]
    [SwaggerOperation(Description = "Редактирование комментария")]
    [ProducesResponseType(StatusCodes.Status200OK)]
    public async Task<CommentEditResponseDto> EditComment([FromQuery] AccName accountName,
        [FromQuery] NotePath pathNote,
        [FromBody] CommentEditRequestDto createDto)
    {
        Check(accountName);

        return await commentEditService.EditContentComment(pathNote, accountName, createDto);
    }

    [HttpDelete]
    [KeycloakAuthorize(Policies.READ_NOTES, Policies.DELETE_COMMENTS)]
    [Route("comments/{pathNote}")]
    [SwaggerOperation(Description = "Удаление комментария")]
    [ProducesResponseType(StatusCodes.Status204NoContent)]
    public async Task<ActionResult> DelComment([FromQuery] AccName accountName,
        [FromQuery] NotePath pathNote)
    {
        Check(accountName);

        await commentEditService.Delete(pathNote, accountName);
        return NoContent();
    }
}

File: javaBack/ResourceServer/src/main/java/com/notes/controllers/PublicCommentsController.java
package com.notes.controllers;

import com.notes.mappers.request.NoteRequestMapper;
import com.notes.mappers.response.NoteResponseMapper;
import com.notes.mappers.response.PageResponseMapper;
import com.notes.models.api.account.AccName;
import com.notes.models.api.note.CommentCreateRequestDto;
import com.notes.models.api.note.CommentCreateResponseDto;
import com.notes.models.api.note.CommentEditRequestDto;
import com.notes.models.api.note.CommentEditResponseDto;
import com.notes.models.api.note.CommentSearchContentResponseDto;
import com.notes.models.api.note.NotePath;
import com.notes.models.api.page.PageDto;
import com.notes.models.api.page.PageLimitDto;
import com.notes.models.api.page.PageSizeDto;
import com.notes.services.managers.CommentEditService;
import com.notes.services.managers.CommentInformationService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestAttribute;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.servlet.support.ServletUriComponentsBuilder;

@RestController
@RequestMapping("/api/public/{accountName}")
@RequiredArgsConstructor
public class PublicCommentsController {
   private final CommentInformationService commentInformationService;
   private final NoteResponseMapper noteResponseMapper;
   private final PageResponseMapper pageResponseMapper;
   private final CommentEditService commentEditService;
   private final NoteRequestMapper noteRequestMapper;

   @GetMapping(path = "/notes/{pathNote}/comments", headers = "content-type=application/json", consumes = MediaType.APPLICATION_JSON_VALUE)
   @PreAuthorize("hasAnyAuthority('READ_COMMENTS')")
   public PageDto<CommentSearchContentResponseDto> comments(
           @Valid @PathVariable AccName accountName,
           @Valid @PathVariable NotePath pathNote,
           @Valid @RequestAttribute PageSizeDto page,
           @Valid @RequestAttribute PageLimitDto limit) {
      var comments = commentInformationService
              .comments(accountName.name(), pathNote.path(), page.page(), limit.limit())
              .map(noteResponseMapper::ofCommentContent);
      return pageResponseMapper.of(comments);
   }

   @PostMapping(path = "/notes/{pathNote}/comments", headers = "content-type=application/json", consumes = MediaType.APPLICATION_JSON_VALUE)
   @PreAuthorize("hasAnyAuthority('READ_COMMENTS', 'READ_NOTES', 'CREATE_COMMENTS')")
   public ResponseEntity<CommentCreateResponseDto> createComment(@Valid @PathVariable AccName accountName,
                                                                 @Valid @PathVariable NotePath pathNote,
                                                                 @Valid @RequestBody CommentCreateRequestDto createDto) {
      var note = commentEditService.createComment(accountName.name(),
              pathNote.path(),
              noteRequestMapper.of(createDto));

      return ResponseEntity.created(ServletUriComponentsBuilder
                      .fromPath("/api/public/{accountName}/notes/{pathNote}")
                      .buildAndExpand(accountName.name(), note.path()).toUri())
              .body(noteResponseMapper.ofCommentCreate(note));
   }

   @PostMapping("/comments/{pathNote}")
   @PreAuthorize("hasAnyAuthority('READ_COMMENTS', 'READ_NOTES', 'EDIT_OWN_COMMENTS')")
   public CommentEditResponseDto editComment(@Valid @PathVariable AccName accountName,
                                             @Valid @PathVariable NotePath pathNote,
                                             @Valid @RequestBody CommentEditRequestDto createDto) {
      var comment = commentEditService.editComment(pathNote.path(),
              accountName.name(),
              noteRequestMapper.of(createDto));

      return noteResponseMapper.ofCommentEdit(comment);
   }

   @DeleteMapping("/comments/{pathNote}")
   @PreAuthorize("hasAnyAuthority('READ_COMMENTS', 'READ_NOTES', 'DELETE_COMMENTS')")
   @ResponseStatus(HttpStatus.NO_CONTENT)
   public void delComment(@Valid @PathVariable AccName accountName,
                          @Valid @PathVariable NotePath pathNote) {
      commentEditService.deleteComment(pathNote.path(), accountName.name());
   }
}
